### PR TITLE
Captions Support

### PR DESCRIPTION
**Pull request by SoneeJohn**
Adds support for retrieving caption URLs from YouTube. Feedback on whether or not this is a good idea is very much appreciated 😊 Also, if for whatever reason this PR cannot be merged, I think an alternative can be achieved by *exposing* the `NSDictionary` that is used during the initialization of `XCDYouTubeVideo` (https://github.com/0xced/XCDYouTubeKit/issues/294)

#118

### CODE REVIEW

**Comment by SoneeJohn** [comment]
@0xced I see that you are using `VCRURLConnection` for network tests. I will update my tests to use this in an coming commit.

**Comment by 0xced** [comment]
Just enable the `VCR_CASSETTES_DIRECTORY` checkbox in the _XCDYouTubeKit iOS Static Library_ scheme in order to automatically record with VCRURLConnection.

**Comment by SoneeJohn** [comment]
@0xced That's neat 🙂! I make the changes in my latest commit. https://github.com/0xced/XCDYouTubeKit/pull/281/commits/68b6b903cd7244d8aacd54e812924d6ca9812a69

**Comment by SoneeJohn** [comment]
@Alexookah I think captions is also a good idea. However, there are some limitations at the moment:

* I only added primitive support therefore, you need to parse the XML document. I created a class that will parse the document and create a SRT file but I don't think @0xced would merge something like that as it might be going beyond the intended purpose of the library.
* You cannot view the captions on a video i.e an AVPlayerView without first processing the video and adding the captions manually to it (if I recall correctly)

**Comment by Alexookah** [comment]
i have allready implemented this adding subtitlies using this library
https://github.com/mhergon/AVPlayerViewController-Subtitles
very easy.

this library parses the .srt files but i simply skipped and parsed the caption data from youtube and it works great.

let subtitlesDictionary = NSMutableDictionary()
if CaptionUrl != nil {
    Alamofire.request(CaptionUrl!).response { response in
        let xml = SWXMLHash.parse(response.data!)
        var counter = 0
        for elem in xml["transcript"]["text"].all {
            let fromTime = Double((elem.element?.attribute(by: "start")?.text)!)
            let toTime = fromTime! + Double((elem.element?.attribute(by: "dur")?.text)!)!
            let text = elem.element?.text
            // Create final object
            let final = NSMutableDictionary()
            final["from"] = fromTime
            final["to"] = toTime
            final["text"] = replacedText
            subtitlesDictionary[counter] = final
            counter += 1
        }
    }
}


// Add subtitles
playerViewController?.addSubtitles().giveString(content: subtitlesDictionary)

// Change text properties
playerViewController?.subtitleLabel?.textColor = UIColor.white


i am sure its not the best way to achieve this.
but once the captions are available i could simply parse them and use this library to add subtitles to the video.

**Comment by SoneeJohn** [comment]
@Alexookah that's a nice library however, I wanted something that was more native to the system.

File: XCDYouTubeKit/XCDYouTubeVideo.m [hunk 1]
```diff
@@ -103,12 +120,13 @@ - (instancetype) initWithIdentifier:(NSString *)identifier info:(NSDictionary *)
...
 
 	NSString *streamMap = info[@"url_encoded_fmt_stream_map"];
+	NSString *captionsMap = info[@"player_response"];
 	NSString *httpLiveStream = info[@"hlsvp"];
 	NSString *adaptiveFormats = info[@"adaptive_fmts"];
 	
 	NSMutableDictionary *userInfo = response.URL ? [@{ NSURLErrorKey: response.URL } mutableCopy] : [NSMutableDictionary new];
 	
-	if (streamMap.length > 0 || httpLiveStream.length > 0)
+	if (streamMap.length > 0 || httpLiveStream.length > 0 || captionsMap.length > 0)
```
Comment (SoneeJohn) [inline XCDYouTubeKit/XCDYouTubeVideo.m]: Should probably remove captionsMap from the comparison.

**Comment by Alexookah** [comment]
how can i test the changes? are they gonna get merged to the released version?

**Comment by SoneeJohn** [comment]
@Alexookah I wrote some unit tests (2) if you would like to explore further then you can clone my fork. In regards to releasing a version I would still like to here @0xced input on this 🙂

**Comment by SoneeJohn** [comment]
@0xced ping! 😝

**Comment by kieuquangloc147** [comment]
Hi @0xced and @SoneeJohn , how is this feature?

**Pull request merged**